// @aws-solutions-constructs/core/lib/vpc-defaults.js
"use strict";
/**
 *  Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultIsolatedVpcProps = exports.DefaultPublicPrivateVpcProps = void 0;
const ec2 = require("@aws-cdk/aws-ec2");
/**
 * Default VPC with public and private subnets
 */
function DefaultPublicPrivateVpcProps() {
    return {};
}
exports.DefaultPublicPrivateVpcProps = DefaultPublicPrivateVpcProps;
/**
 * Default VPC with isolated subnets
 */
function DefaultIsolatedVpcProps() {
    return {
        natGateways: 0,
        subnetConfiguration: [
            {
                cidrMask: 18,
                name: "isolated",
                subnetType: ec2.SubnetType.ISOLATED,
            }
        ]
    };
}
exports.DefaultIsolatedVpcProps = DefaultIsolatedVpcProps;
